extension_distribution.md

```diff
@@ -0,0 +1,296 @@
+analogue is Helm: right now we're focused on the analgue for Helm repos, not the
+analogue for ArtifactHub or CharmHub.
+
+### Tiltfile 
+
+We see the future of Tilt extensibility as the API Server and Kubernetes
+operator pattern.
+
+What that means: there's an HTTP server that allows you can read, watch, and
+write data models. Long-term, to add a new feature to Tilt, you will add a new
+resource/controller pair.
+
+The Tiltfile is a client of the API Server. We do NOT think that adding more
+hooks and extension points to the Tiltfile execution model is the right answer.
+
+## Prior Art
+
+### Other Extension Systems
+
+Some other extension systems that we looked to for inspiration:
+
+#### VSCode extensions (https://code.visualstudio.com/docs/editor/extension-marketplace)
+
+You install them into VSCode itself (rather than a particular repo).
+
+They typically add things to the Command Palette. But there are also extensions
+that manage other tools on your machine (e.g., the Minikube extension makes sure
+Minikube is up to date).
+
+#### Browser extensions
+
+These typically come in two flavors:
+
+- Extensions that add buttons or capapilities to the browser itself (like [tabby
+  cat](https://chrome.google.com/webstore/detail/tabby-cat/mefhakmgclhhfbdadeojlkbllmecialg?hl=en),
+  the best extension)
+
+- Extensions that modify pages you visit, like Google Translate.
+
+#### Gradle Plugins (https://docs.gradle.org/current/userguide/implementing_gradle_plugins.html)
+
+Add custom task types to your build system, or decorate existing tasks with new
+outputs. Here are some examples: https://github.com/ksoichiro/awesome-gradle
+
+#### Bazel Extensions (https://docs.bazel.build/versions/main/skylark/concepts.html)
+
+Mostly focused on adding macros and custom build rules.
+
+#### Helm Chart Repositories (https://helm.sh/docs/helm/helm_repo/)
+
+The Helm repo command lets you install new repositories of charts. Each chart is a set of resources
+to deploy. These charts can add new types and operators to your Kubernetes cluster.
+
+The chart repos are globally installed, but the charts are installed to a particular cluster.
+
+#### yarn add / yarn global add
+
+In the javascript ecoysystem, packages are installed locally (in the current
+repo) or globally (in a shared global package tracker). Packages can contain
+both library code that you have to explicitly import, but also contains
+executables that you can run.
+
+## Proposal
+
+We'll add three new data types to the APIServer:
+
+- `ExtensionRepo`: A repository of Tilt extensions.
+- `Extension`: A globally installed extension.
+- `Tiltfile`: A Tiltfile invocation.
+
+Most data types in the Tilt APIServer are stored in-memory only, and disappear
+when Tilt is killed. The `ExtensionRepo` and `Extension` types are
+different. They're persisted on-disk across Tilt sessions under ~/.tilt-dev.
+
+### ExtensionRepo
+
+```go
+type ExtensionRepoSpec struct {
```

Comment:
I wonder if it'd make sense to combine `ExtensionRepo` + `Extension`. My main concern is that if we expect the primary way for extensions to be consumed is via Git repo, versioning could become complicated.

Being able to e.g. use `go-getter` type syntax and have something like `github.com/my-company/my-extensions//extension-1?ref=abc123` would be nice.

(A full "repo" is more useful IMO in the Helm type model where there's some sort of index that lives there.)

Reply:
I could probably be convinced that we don't need both Extension + ExtensionRepo.

My overall thinking here:

Pros:
- Adding repos now is relatively cheap
- in every package system I've ever seen, they had a some notion of a repo. And if they didn't have one from the beginning, it was extremely painful to add later (think Golang)
- discoverability and indexes will be much easier to add later

Cons:
- over-engineering
- an extra step to interact with
- we might end up having to change ExtensionRepo later

I see the versioning situation as a mixed bag -- if you assume each org has their own extension repo, I could imagine big benefits to forcing you to version them together (think: Brew Taps). Though I agree that version combinatorial explosions have upsides and downsides.

Reply:
I definitely agree with your points, particularly about the importance of repos as a concept.

I do kind of like the way Python/npm+yarn handle things, where an unbound name refers to something from a repo and URL means go grab it directly.

Perhaps Helm is actually a better model here: `helm install mynginx foo/nginx` will expect to get `nginx` chart from `foo` repo while `helm install mynginx https://example.com/charts/nginx-1.2.3.tgz`will grab it by URL directly.

I guess model-wise, we could always support syntactic sugar for those use cases and just have a 1:1 repo:extension relationship behind the scenes.

And I definitely like the idea of a Git repo being able to serve _as_ a repository since I imagine a high-percentage of extensions users will want to keep private, and realistically that's the lowest friction way in any enterprise for a dev team to spin up new "infrastructure" 😂 

Also, re: versioning, I guess there's nothing to stop us from just checking out the same repo @ different refs per-extension if we wanted to support that in the future (I wonder if `git-go` supports [`git-worktree`](https://git-scm.com/docs/git-worktree) 🤔 )

soooo tl;dr I'm totally fine either way 😛

Reply:
:shrug: 

i'll probably end up playing with both a bit; i don't think there's a right answer here